partition/ubootenv/env.go

```diff
@@ -0,0 +1,291 @@
+	f, err := os.Create(fname)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	env := &Env{
+		fname: fname,
+		size:  size,
+		data:  make(map[string]string),
+	}
+
+	return env, nil
+}
+
+// OpenFlags instructs open how to alter its behavior.
+type OpenFlags int
+
+const (
+	// OpenBestEffort instructs OpenWithFlags to skip malformed data without returning an error.
+	OpenBestEffort OpenFlags = 1 << iota
+)
+
+// Open opens a existing uboot env file
+func Open(fname string) (*Env, error) {
+	return OpenWithFlags(fname, OpenFlags(0))
+}
+
+// OpenWithFlags opens a existing uboot env file, passing additional flags.
+func OpenWithFlags(fname string, flags OpenFlags) (*Env, error) {
+	f, err := os.Open(fname)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	contentWithHeader, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+	crc := readUint32(contentWithHeader)
+
+	payload := contentWithHeader[headerSize:]
+	actualCRC := crc32.ChecksumIEEE(payload)
+	if crc != actualCRC {
+		return nil, fmt.Errorf("bad CRC: %v != %v", crc, actualCRC)
+	}
+	eof := bytes.Index(payload, []byte{0, 0})
+
+	data, err := parseData(payload[:eof], flags)
+	if err != nil {
+		return nil, err
+	}
+
+	env := &Env{
+		fname: fname,
+		size:  len(contentWithHeader),
+		data:  data,
+	}
+
+	return env, nil
+}
+
+func parseData(data []byte, flags OpenFlags) (map[string]string, error) {
+	out := make(map[string]string)
+
+	for _, envStr := range bytes.Split(data, []byte{0}) {
+		if len(envStr) == 0 || envStr[0] == 0 || envStr[0] == 255 {
+			continue
+		}
+		l := strings.SplitN(string(envStr), "=", 2)
+		if len(l) != 2 {
+			if flags&OpenBestEffort == OpenBestEffort {
+				continue
+			}
+			return nil, fmt.Errorf("cannot parse line %q as key=value pair", envStr)
+		}
+		key := l[0]
+		value := l[1]
+		out[key] = value
+	}
+
+	return out, nil
+}
+
+func (env *Env) String() string {
+	out := ""
+
+	env.iterEnv(func(key, value string) {
+		out += fmt.Sprintf("%s=%s\n", key, value)
+	})
+
+	return out
+}
+
+func (env *Env) Size() int {
+	return env.size
+}
+
+// Get the value of the environment variable
+func (env *Env) Get(name string) string {
+	return env.data[name]
+}
+
+// Set an environment name to the given value, if the value is empty
+// the variable will be removed from the environment
+func (env *Env) Set(name, value string) {
+	if name == "" {
+		panic(fmt.Sprintf("Set() can not be called with empty key for value: %q", value))
+	}
+	if value == "" {
+		delete(env.data, name)
+		return
+	}
+	env.data[name] = value
+}
+
+// iterEnv calls the passed function f with key, value for environment
+// vars. The order is guaranteed (unlike just iterating over the map)
+func (env *Env) iterEnv(f func(key, value string)) {
+	keys := make([]string, 0, len(env.data))
+	for k := range env.data {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	for _, k := range keys {
+		if k == "" {
+			panic("iterEnv iterating over a empty key")
```

Comment:
this is ok as long as make sure keys in data must never be empty by construction